Bulk Access Control Script REST implementation

eskander17 opened the pull request:
## References
* More details about the whole proposal and wireframe can be found there https://wiki.lyrasis.org/display/DSPACE/Bulk+access+control+management+-+4Science+proposal
* Fixes https://github.com/DSpace/DSpace/issues/2848
* Related to [DSpace/RestContract#223](https://github.com/DSpace/RestContract/pull/223)
* Requires Frontend PR: https://github.com/DSpace/dspace-angular/pull/2247

Please note that this PR is built on top of https://github.com/DSpace/DSpace/pull/8830

## Description
1. A new endpoint has been introduced to expose the configuration available for use with the bulk access control features (list of available access conditions at item and bitstream level)
2. a BulkAccessControl Script available to admins of target item has been introduced to perform bulk changes over the access condition of the selected objects

## Instructions for Reviewers
You can use the corresponding angular PR to be linked with to test the currect behavior of the script.
Take a look to the documentation available there https://wiki.lyrasis.org/display/DSPACE/Bulk+access+control+management+-+4Science+proposal
in short you have access to the new features
- as a repository administrator from the left menu inside the access control section
- as a local administrator from any objects (community, collection, item) that you administer via the administer page, a dedicated tab has been added

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon. Please note that the PR is smaller than what it looks as it is built on top of another PR and there are large ITs
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [n/a] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [X] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [X] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

Commit:
[CST-9634] created new endpoint for bulkAccessConditionOptions configuration

eskander17 committed:
[CST-9634] refactoring

Commit:
bulk access controll

Commit:
refactoring BulkAccessControl

abollini committed:
CST-9634 fix issues with date timezone and inheritance of default policies

eskander17 committed:
refactoring and adding new test cases

eskander17 committed:
Merge remote-tracking branch 'origin/duracom-136_CST-9634' into duracom-136_CST-9634

# Conflicts:
#	dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControl.java
#	dspace-api/src/test/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlIT.java

eskander17 committed:
refactoring and adding new test cases

atarix83 committed:
[CST-9634] Fix exception message

atarix83 committed:
[CST-9634] add administrativeBulkAccess discovery configuration

Commit:
refactoring and added new IT class

eskander17 committed:
Merge remote-tracking branch 'origin/duracom-136_CST-9634' into duracom-136_CST-9634

tdonohue commented:
@benbosman : I'm adding you as a reviewer of this major feature, but feel free to delegate it to someone else on your team.  I'd just appreciate it if Atmire can help with the review.

atarix83 committed:
Merge branch 'main' into duracom-136_CST-9634

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControl.java:
```diff
@@ -0,0 +1,439 @@
+        assignCurrentUserInContext();
+        assignSpecialGroupsInContext();
+
+        context.turnOffAuthorisationSystem();
```
This seems dangerous.  We appear to **always** run the bulk access control without any permissions checks.  Why aren't we using the current user & special groups just before this?

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlScriptConfiguration.java:
```diff
@@ -0,0 +1,110 @@
+        if (options == null) {
+            Options options = new Options();
+
+            options.addOption("u", "target", true, "target uuids of communities/collections/items");
```
I find this odd that we are mapping `--target` to `-u`.  I would expect either the long form should be renamed `--uuids` or we should map it to be `-t`.  Mapping one letter to another seems odd.

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlScriptConfiguration.java:
```diff
@@ -0,0 +1,110 @@
+            options.getOption("f").setType(InputStream.class);
+            options.getOption("f").setRequired(true);
+
+            options.addOption("h", "help", false, "help");
```
This param doesn't appear to work.  From command-line, I get an error if I try to use `-h`: 
```
 # ./dspace bulk-access-control -h
The script has started
A target uuid must be provided with at least on uuid (run with -h flag for details)
java.lang.IllegalArgumentException: At least one target uuid must be provided
        at org.dspace.app.bulkaccesscontrol.BulkAccessControl.internalRun(BulkAccessControl.java:137)
        at org.dspace.scripts.DSpaceRunnable.run(DSpaceRunnable.java:104)
        at org.dspace.app.launcher.ScriptLauncher.executeScript(ScriptLauncher.java:149)
        at org.dspace.app.launcher.ScriptLauncher.handleScript(ScriptLauncher.java:131)
        at org.dspace.app.launcher.ScriptLauncher.main(ScriptLauncher.java:98)
```

eskander17 replied:
Done

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlScriptConfiguration.java:
```diff
@@ -0,0 +1,110 @@
+
+            options.addOption("h", "help", false, "help");
+
+            super.options = options;
```
I also noticed there's no way to specify the user to run this as via commandline.  The script always ignores all permissions and doesn't require a user to be logged in.

Ideally, via commandline there should be a **required** user param (`--eperson` or `-e`).   But, when run from the UI, then it should just use the currently logged in user and *not* allow for the `--eperson` param to be specified.

For example see the `./dspace import` script which acts in the manner.  I believe a few other scripts also work the same way.

eskander17 replied:
Done

tdonohue reviewed (changes requested):
@eskander17 : Thanks!  Overall, this appears to be working and the direction is a good one.  But, I've found some issues with the code (the most severe is that it ignores all permissions when executed).   There's also some minor cleanup that I've noted inline below.

I also want to point out this appears to be currently failing some Integration Tests.  Those need to be fixed before we can move this forward.

I have not noticed any obvious bugs, which is good.  Just some code issues.

tdonohue commented:
@eskander17 : I found one additional issue with more testing.  It appears this script never has any useful output when it succeeds (either from UI or commandline).  Here's a UI screenshot to show what I mean:

![no-output](https://github.com/DSpace/DSpace/assets/483997/74bb67b6-f675-4633-8300-1e884011d9bb)

I'd expect it to output which objects were updated (since it's a bulk tool) and ideally list what updates were made.  Something similar to this sort of output:
```
Replacing Item {uuid} policy to access condition: "lease", end_date=2024-01-02
...
Adding Bitstream {uuid} policy with access condition: "administer"
...
Replacing Item {uuid} policies to inherit from owning Collection.
```

This information should be relatively easy to log and it'd be an extremely useful report of what changes were made in the system.

abollini commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControl.java:
```diff
@@ -0,0 +1,439 @@
+    private void updateBitstreamsPolicies(Item item, AccessControl accessControl) {
+        AccessConditionBitstream.Constraint constraints = accessControl.getBitstream().getConstraints();
+
+        item.getBundles(CONTENT_BUNDLE_NAME).stream()
```
@tdonohue you have actually identified a bug / inaccurate implementation in respect to the original requirement. We are going to fix it adding dedicated test to proof the changes.
Please check the corresponding comment on the rest contract https://github.com/DSpace/RestContract/pull/223#discussion_r1209463843

abollini commented on dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java:
```diff
@@ -940,14 +940,39 @@ public void adjustBundleBitstreamPolicies(Context context, Item item, Collection
+        List<ResourcePolicy> defaultItemPolicies = authorizeService.findPoliciesByDSOAndType(context, item,
+                ResourcePolicy.TYPE_CUSTOM);
+        if (defaultCollectionPolicies.size() < 1) {
+            throw new SQLException("Collection " + collection.getID()
```
This code is only the result of a refactoring so that we can apply the same logic to a single bitstream or to all the bitstreams in an item, see the existing method in the main branch
https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/content/ItemServiceImpl.java#L922

Collections are required to have a DEFAULT_ITEM_READ and DEFAULT_BITSTREAM_READ policies, this is a constraint defined long time ago.

abollini commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/BulkAccessConditionRestRepository.java:
```diff
@@ -0,0 +1,84 @@
+    private AuthorizeService authorizeService;
+
+    @Override
+    @PreAuthorize("permitAll()")
```
The method check the permissions at the very beginning. If you like we can move the isAuthorized method implementation in a specific Bean and use it in the PreAuthorize annotation via the SpEL. Similarly to what is done for the versioning feature, see https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/security/VersioningSecurityBean.java

tdonohue replied:
@abollini : Yes, I saw that `isAuthorized(context)` is called at the beginning of this method.  It just seemed odd to me that nothing *first* verifies there's an authenticated user.

Currently, `isAuthorized(context)` looks like this:
```
return authorizeService.isAdmin(context) || authorizeService.isComColAdmin(context) ||
                authorizeService.isItemAdmin(context);
```

This means that if an unauthenticated user calls this endpoint, they will go through all 3 of these checks... and all three will obviously return false.  So it seems odd not to surround this all with an initial check to verify whether there's even an authenticated user **before** checking each of these permission levels.

eskander17 replied:
Done

abollini commented on dspace-server-webapp/src/test/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlScriptIT.java:
```diff
@@ -0,0 +1,502 @@
+                    .param("properties", new ObjectMapper().writeValueAsString(list)))
+            .andExpect(status().isForbidden());
+    }
+
```
I agree, it is required to add IT to proof that the security is respected. It is via the REST API/UI as we have tried it but in any case it is better to enforce with a test.
Please note that the script needs to be visible to any community/collection/item administrator, and this is what is tested in the "if" here
https://github.com/DSpace/DSpace/pull/8844/files#diff-4c8c1821648640a4c6f4a941b1f67a841f549bbdbe2f0e2217a10ecb6acaf12bR41-R44

when the script is actually executed we will go in the "else" branch where specific permission over the target are verified
https://github.com/DSpace/DSpace/pull/8844/files#diff-4c8c1821648640a4c6f4a941b1f67a841f549bbdbe2f0e2217a10ecb6acaf12bR45-R65

tdonohue replied:
It appears that the extra permission checks are still missing in our ITs.  I don't see the following:

1. Verify that a Community Admin cannot run this script on a different Community (or a Collection not in their Community)
2. Verify that a Collection Admin cannot run this script on a different Collection (or Community)
3. Verify that an Item Admin cannot run this script on a different Item (or Community or Collection).

The only new test I see related to these checks is in `BulkAccessControlIT.performBulkAccessWithNotCollectionAdminEPersonTest()`.  This is a decent test for the second point... but the first and third are still missing in our tests.

Commit:
- refactoring and adding new test cases
- update resource polices of derivative bitstreams also
- check authorization when run script
- added some log information

Commit:
added missed configuration

eskander17 committed:
Merge remote-tracking branch 'origin/main' into duracom-136_CST-9634

Commit:
edited param name

abollini committed:
CST-9634 Assure that REST API assumes all dates to be UTC if not specified

abollini committed:
Merge branch 'main' of https://github.com/DSpace/DSpace into duracom-136_CST-9634

abollini committed:
Merge branch 'duracom-136_CST-9634' of https://bitbucket.org/4Science/dspace-cris into duracom-136_CST-9634

abollini commented:
The test failures that was reported up to now were related to bugs solved in https://github.com/DSpace/DSpace/pull/8698 this PR has made visible the bug as we have forced H2 to store date in UTC that usually is not the timezone of the VM used by the CI.

tdonohue commented:
@abollini : It appears this still has CI build failures, but now there's a compilation failure:
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project dspace-api: Compilation failure
Error:  /home/runner/work/DSpace/DSpace/dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControl.java:[280,63] error: validateResourcePolicy(Context,String,Date,Date) has private access in AccessConditionOption
```

abollini committed:
CST-9634 make public the validateResourcePolicy method

abollini committed:
CST-9634 remove unnecessary turn off/on auth system

abollini committed:
Merge branch 'main' of https://github.com/DSpace/DSpace into duracom-136_CST-9634

abollini reviewed (approved):
@tdonohue your previous comments seem to be addressed. The code looks good to me and functional. As you can see from the commit history I have directly supported the resolution of some test issue mostly not related with the main development but in any case it would be better to have an independent check

eskander17 commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/model/AccessConditionBitstream.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * @author Mohamed Eskander (mohamed.eskander at 4science.it)
+ */
+public class AccessConditionBitstream {
```
Done

eskander17 commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/model/AccessConditionItem.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * @author Mohamed Eskander (mohamed.eskander at 4science.it)
+ */
+public class AccessConditionItem {
```
Done

eskander17 commented on dspace-api/src/main/java/org/dspace/content/service/ItemService.java:
```diff
@@ -507,6 +507,25 @@ public void inheritCollectionDefaultPolicies(Context context, Item item, Collect
+     * @throws SQLException       If database error
+     * @throws AuthorizeException If authorization error
+     */
+    public void adjustBitstreamPolicies(Context context, Item item, Collection collection, Bitstream bitstream)
```
Done

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControlCliScriptConfiguration.java:
```diff
@@ -0,0 +1,42 @@
+        options.getOption("f").setType(InputStream.class);
+        options.getOption("f").setRequired(true);
+
+        options.addOption("e", "eperson", true, "email of eperson doing importing");
```
Minor thing. This description is incorrect. It should likely say "email of EPerson used to perform actions" (or similar)

eskander17 replied:
Done

tdonohue commented on dspace-api/src/main/java/org/dspace/app/bulkaccesscontrol/BulkAccessControl.java:
```diff
@@ -0,0 +1,538 @@
+        }
+    }
+
+    private void validateAccessCondition(AccessCondition accessCondition) {
```
This entire class includes all the core code behind this feature, yet very few of the methods have any description or code comments.  Could we please add JavaDocs to these private methods to describe the purpose of each.  I know you tried to name them in a manner that "makes sense", but these methods could still use more explanation.  For example, what makes an access condition "valid"?  And we should document that we throw a BulkAccessControlException if invalid.

The same goes for most of the methods in this class (except maybe simple getters or logging).  They should each have basic descriptions because these methods perform most of the actions.

eskander17 replied:
Done

tdonohue reviewed (changes requested):
@eskander17 and @abollini :  This is looking better overall, but I still have some inline feedback and we have ITs that are still missing.

I've also found that I can no longer seem to run a bulk access change on all Bitstreams in a Collection or Community.  This fails from both the UI and the commandline with this error:
```
org.dspace.content.Bundle.bitstreams, could not initialize proxy - no Session
	at org.hibernate.collection.internal.AbstractPersistentCollection.throwLazyInitializationException(AbstractPersistentCollection.java:614)
	at org.hibernate.collection.internal.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:218)
	at org.hibernate.collection.internal.AbstractPersistentCollection.initialize(AbstractPersistentCollection.java:591)
	at org.hibernate.collection.internal.AbstractPersistentCollection.read(AbstractPersistentCollection.java:149)
	at org.hibernate.collection.internal.PersistentList.toArray(PersistentList.java:161)
	at java.base/java.util.ArrayList.<init>(ArrayList.java:179)
	at org.dspace.content.Bundle.getBitstreams(Bundle.java:140)
	at org.dspace.core.HibernateDBConnection.uncacheEntity(HibernateDBConnection.java:294)
	at org.dspace.core.HibernateDBConnection.uncacheEntity(HibernateDBConnection.java:287)
	at org.dspace.core.Context.uncacheEntity(Context.java:880)
	at org.dspace.app.bulkaccesscontrol.BulkAccessControl.updateItemsAndBitstreamsPolices(BulkAccessControl.java:309)
	at org.dspace.app.bulkaccesscontrol.BulkAccessControl.internalRun(BulkAccessControl.java:183)
	at org.dspace.scripts.DSpaceRunnable.run(DSpaceRunnable.java:104)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

To reproduce, you can attempt to change all Bitstreams to "administrator" access rights in a Community or Collection.  Here's an example `data.json` to use from commandline:
```
{
  "bitstream": {
    "constraints": {
      "uuid": []
    },
    "mode": "replace",
    "accessConditions": [
      {
        "name": "administrator",
        "startDate": null,
        "endDate": null
      }
    ]
  }
}
```

Beyond that, other tests that I've performed are working.  So, this is looking good overall, but some more issues to fix/cleanup.

Commit:
added new test cases
added JavaDocs
refactoring

eskander17 committed:
Merge remote-tracking branch 'origin/main' into duracom-136_CST-9634

# Conflicts:
#	dspace/config/spring/api/discovery.xml

Commit:
solved issue of 'could not initialize proxy - no Session'

eskander17 commented:
> @eskander17 and @abollini : This is looking better overall, but I still have some inline feedback and we have ITs that are still missing.
> 
> I've also found that I can no longer seem to run a bulk access change on all Bitstreams in a Collection or Community. This fails from both the UI and the commandline with this error:
> 
> ```
> org.dspace.content.Bundle.bitstreams, could not initialize proxy - no Session
> 	at org.hibernate.collection.internal.AbstractPersistentCollection.throwLazyInitializationException(AbstractPersistentCollection.java:614)
> 	at org.hibernate.collection.internal.AbstractPersistentCollection.withTemporarySessionIfNeeded(AbstractPersistentCollection.java:218)
> 	at org.hibernate.collection.internal.AbstractPersistentCollection.initialize(AbstractPersistentCollection.java:591)
> 	at org.hibernate.collection.internal.AbstractPersistentCollection.read(AbstractPersistentCollection.java:149)
> 	at org.hibernate.collection.internal.PersistentList.toArray(PersistentList.java:161)
> 	at java.base/java.util.ArrayList.<init>(ArrayList.java:179)
> 	at org.dspace.content.Bundle.getBitstreams(Bundle.java:140)
> 	at org.dspace.core.HibernateDBConnection.uncacheEntity(HibernateDBConnection.java:294)
> 	at org.dspace.core.HibernateDBConnection.uncacheEntity(HibernateDBConnection.java:287)
> 	at org.dspace.core.Context.uncacheEntity(Context.java:880)
> 	at org.dspace.app.bulkaccesscontrol.BulkAccessControl.updateItemsAndBitstreamsPolices(BulkAccessControl.java:309)
> 	at org.dspace.app.bulkaccesscontrol.BulkAccessControl.internalRun(BulkAccessControl.java:183)
> 	at org.dspace.scripts.DSpaceRunnable.run(DSpaceRunnable.java:104)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> ```
> 
> To reproduce, you can attempt to change all Bitstreams to "administrator" access rights in a Community or Collection. Here's an example `data.json` to use from commandline:
> 
> ```
> {
>   "bitstream": {
>     "constraints": {
>       "uuid": []
>     },
>     "mode": "replace",
>     "accessConditions": [
>       {
>         "name": "administrator",
>         "startDate": null,
>         "endDate": null
>       }
>     ]
>   }
> }
> ```
> 
> Beyond that, other tests that I've performed are working. So, this is looking good overall, but some more issues to fix/cleanup.

@tdonohue Done

tdonohue reviewed (approved):
👍 Thanks @eskander17 !  I re-tested this today and the prior bugs were fixed.  I'm not finding any new issues in my testing and the code changes look good.   So, I feel  this is ready to merge as soon as the frontend PR is ready.

@benbosman : could you or someone on your team still give this a test (at a minimum)?  This is a very large, complex feature, so additional testers would be appreciated.

tdonohue commented:
@eskander17 and @abollini : The merge conflicts with this PR were the result of merging #8830 a moment ago.  If the conflicts could be resolved quickly, it'd be helpful to reviewers.  Thanks!

abollini committed:
Merge branch 'main' of https://github.com/DSpace/DSpace into duracom-136_CST-9634

abollini commented:
@tdonohue merge conflict solved

tdonohue reviewed (approved):
👍 Thanks @eskander17 and @abollini .  Re-tested today after merge conflict resolution.  No issues found.  This looks good to me.

tdonohue commented:
@benbosman : Did you still want to give this a quick test?  I'd like to get this merged tomorrow (Thurs, June 15) if possible to begin final testing of the 7.6 release... I was waiting on your feedback as a secondary tester because this PR is rather large and could use more testers.   As @artlowel tested the frontend PR, that may be "good enough" though, assuming you agree.

tdonohue commented:
As discussed in today's Dev Meeting, I will merge this tomorrow (Friday).  Leaving this open for one last day, in case @benbosman has a chance to review/test it.  After that, we'll assume the tests from myself, @artlowel (who tested the frontend) and the 4Science team are sufficient -- we feel they likely are, but more eyes are always welcome on large features.
